railties/lib/rails/generators/rails/app/templates/.devcontainer/devcontainer.json.tt

```diff
@@ -0,0 +1,34 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/ruby
+{
+  "name": "<%= app_name %>",
+  "dockerComposeFile": "docker-compose.yml",
+  "service": "rails-app",
+  "workspaceFolder": "/workspaces/${localWorkspaceFolderBasename}",
+
+  // Features to add to the dev container. More info: https://containers.dev/features.
+  "features": {
+    "ghcr.io/devcontainers/features/github-cli:1": {}
+  },
+
+<%- if !devcontainer_variables.empty? -%>
+  "containerEnv": {
+    <%= devcontainer_variables.map { |key, value| "\"#{key}\": \"#{value}\"" }.join(",\n    ") %>
+  },
+<%- end -%>
```

Comment:
Adding it to compose file directly, e.g., as an extension, would make it possible to use the same compose file without Dev Containers.

```yml
x-env: &app-env
  DB_HOST: postgres
  REDIS_URL: redis://redis:6379/1

services:
      rails-app:
        environment:
          <<: *app-env

```

I know, it is not in the scope of this PR to support `docker compose -f .devcontainer/compose.yaml up -d`, but this small difference would make it possible.